File: structures/modifyStructures/py/shiftData/shiftData.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Wed Nov  7 10:13:06 2018

@author: ty

mave atoms in Si/Ge interfaces that have drifted to the other side during 
relaxation back to thier original positions

read in the atomic positions. find the ge atoms in the 'leftmost' part and
cut them from the structures. shift the structure to xmin = 0.

find several intermonolayer distances on the right side, i.e. 50% of Ge region.
average the distances and append the ge atoms cut from the si side to the ge
region with the calculated monolayer distance. Write to file.

'data.relaxed' has atom positions followed some other numbers that 
dont matter, and then all atomic velocities. Only read in positions. Velocities
are 're-seeded' during NEMD.

"""

import numpy as np
import copy as cp

infile = 'data.relaxed'
yz = 8 #no of unit cells in y,z direction

no = yz*yz*2 #no. atoms per monolayer

### READ DATA ###
with open(infile, 'r') as fid:
    fid.readline() #skip comments
    fid.readline()
    
    size = int(fid.readline().strip().split()[0]) #Number of atoms
    fid.readline()
    fid.readline()
    
    tmp = fid.readline().strip().split()
    xlo = float(tmp[0])
    xhi = float(tmp[1])
    tmp = fid.readline().strip().split()
    ylo = float(tmp[0])
    yhi = float(tmp[1])
    tmp = fid.readline().strip().split()
    zlo = float(tmp[0])
    zhi = float(tmp[1])
    
    for i in range(8):
        fid.readline() #skip the rest of the comments
    
    data = np.zeros((size,5)) #original positions
    for i in range(size):
        tmp = fid.readline().strip().split()
        data[i,0] = int(tmp[0]) #id
        data[i,1] = int(tmp[1]) #type
        data[i,2] = float(tmp[2]) #x
        data[i,3] = float(tmp[3]) #y
        data[i,4] = float(tmp[4]) #z
        
### SHIFT GE ATOMS ###
data = data[data[:,2].argsort(),:] #sort by x coord

ids = np.argwhere(data[0:(size/3),1] == 2) #Ge atoms in first 3rd of structure
cut = data[ids[:,0],:] #Ge atoms
data = np.delete(data,ids[:,0],axis=0) #cut Ge atoms from structure

dmin = data[:,2].min() #minimum x coord of structure
data[:,2] = data[:,2]-dmin
nsize = len(data) #size of new array
dmax = data[nsize-no:nsize,2].mean() #average coord of last monolayer

dist = 0 #avg interlayer distance
for i in range(nsize/8/no):
    lo1 = nsize-(i+2)*128
    hi1 = nsize-(i+1)*128
    lo2 = nsize-(i+1)*128
    hi2 = nsize-i*128
    tmp = data[lo2:hi2,2].mean()-data[lo1:hi1,2].mean() 
    dist = dist+tmp
    
dist = dist/(nsize/8/no) #avg interlayer distance

del hi1, hi2, lo1, lo2, tmp #clean up variables

if len(cut) > yz*yz/2: #avg of lo coord of ge atoms
    cmin = cut[0:yz*yz/2,2].mean()
else:
    cmin = cut[:,2].mean()
    
if cmin < 0:
    cut[:,2] = cut[:,2]+abs(cmin)+dmax+dist #shift ge atoms
else:
    cut[:,2] = cut[:,2]-abs(cmin)+dmax+dist #shift ge atoms
    
data = np.append(data,cut,axis=0)
data = data[np.lexsort((data[:,4],data[:,3],data[:,2]))] #sort data

### WRITE TO FILE ###
xlo = data[:,2].min()
xhi = data[:,2].max()
vac = 20 #len of vaccuum to add to each end

with open('EXAMPLEshifted.xyz', 'w') as fid: #xyz file to view structure
    fid.write(str(size)+'\n')
    fid.write('comments\n')
    for i in range(size-1):
        if data[i,1] == 1:
            fid.write('Si\t'+str(data[i,2])+'\t'+str(data[i,3])
            +'\t'+str(data[i,4])+'\n')
        if data[i,1] == 2:
            fid.write('Ge\t'+str(data[i,2])+'\t'+str(data[i,3])
            +'\t'+str(data[i,4])+'\n')
    if data[i,1] == 1:
            fid.write('Si\t'+str(data[i,2])+'\t'+str(data[i,3])
            +'\t'+str(data[i,4]))
    if data[i,1] == 2:
            fid.write('Ge\t'+str(data[i,2])+'\t'+str(data[i,3])
            +'\t'+str(data[i,4]))

with open('data.shifted', 'w') as fid: #lammps data file
    fid.write(str('LAMMPS DATA FILE\n\n'))
    fid.write(str(size) + ' atoms\n')
    fid.write('2 atom types\n\n')
    fid.write(str(xlo-vac)+' '+str(xhi+vac)+' xlo'+' xhi\n')
    fid.write(str(ylo)+' '+str(yhi)+' ylo'+' yhi\n')
    fid.write(str(zlo)+' '+str(zhi)+' zlo'+' zhi\n\n')
    fid.write('Masses\n\n')
    fid.write('1 28.0855\n2 72.64\n\n')
    fid.write('Atoms\n\n')
    for i in range(size-1):
        fid.write(str(int(i+1))+' '+str(int(data[i,1]))+' '+str(data[i,2])
        +' '+str(data[i,3])+' '+str(data[i,4])+'\n')
    fid.write(str(size)+' '+str(int(data[-1,1]))+' '+str(data[-1,2])+' '+
              str(data[-1,3])+' '+str(data[-1,4]))






    









 

File: structures/modifyStructures/matlab/shift_data/shift_data.m
clear all;

fid = fopen('data.relaxed');

tmp = textscan(fid, '%d%s', 1, 'CommentStyle', 'LAMMPS');
sys = tmp{1,1};

tmp = textscan(fid, '%s', 16);
tmp = textscan(fid, '%d%f', 2);
masses = tmp{1,2};

tmp = textscan(fid, '%s', 3);
tmp = textscan(fid, '%d%d%f%f%f%d%d%d', sys);

fclose(fid);

ids = double(tmp{1,1});
types = double(tmp{1,2});
x = tmp{1,3};
y = tmp{1,4};
z = tmp{1,5};
Pos = [ids, types, x, y, z];

[~,sort] = sort(Pos(:,3));
Pos = Pos(sort,:);

len = size(Pos,1)*(1/4);
ids = [];

for i = 1:len
    if Pos(i,2) == 2
        ids = [ids; i];
    end
end
        
block = Pos(ids,:);
Pos(ids,:) = [];

x_shift = (mean(Pos(end-128+1:end,3)) - mean(Pos(end-2*128+1:end-128,3)));
disp('x_shift =');
disp(x_shift);

x_min = min(Pos(:,3));
if x_min <= 0
	Pos(:,3) = (Pos(:,3) + abs(x_min));
else
	Pos(:,3) = (Pos(:,3) - abs(x_min));
end
disp('Pos min =');
disp(min(Pos(:,3)));
disp('Pos max =');
disp(max(Pos(:,3)));

block_min = mean(block(1:48,3));
%block_min = min(block(:,3));
if block_min <= 0
	block(:,3) = (block(:,3) + abs(block_min));
else
	block(:,3) = (block(:,3) - abs(block_min));
end
disp('Block min =');
disp(min(block(:,3)));

x_max = mean(Pos(end-128+1:end,3));
disp('Pos max =');
disp(x_max);

block(:,3) = (block(:,3) + x_max + x_shift);

Pos = [Pos; block];

%%
buffer = 0.009;

xlo = min(Pos(:,3))-buffer-20; xhi = max(Pos(:,3))+buffer+20;
ylo = min(Pos(:,4))-buffer; yhi = max(Pos(:,4))+buffer;
zlo = min(Pos(:,5))-buffer; zhi = max(Pos(:,5))+buffer;

%% Build the Header
fid = fopen(sprintf('data.%d.%s',sys,'shifted'),'w');
fprintf(fid,'LAMMPS data file\r\n\r\n%d atoms\r\n\r\n%d atom types\r\n\r\n'...
    ,sys,length(masses));
fprintf(fid,'%6.3f %6.3f xlo xhi\r\n%6.3f %6.3f ylo yhi\r\n%6.3f %6.3f zlo zhi\r\n'...
    ,xlo,xhi,ylo,yhi,zlo,zhi);
fprintf(fid,'\r\nMasses\r\n\r\n');
for i = 1:length(masses)
    fprintf(fid,'%d %6.4f\r\n',i,masses(i));
end
fprintf(fid,'\r\nAtoms\r\n\r\n');
%% Atoms
for i = 1:sys  
    fprintf(fid,'%d\t%d\t%6.3f\t%6.3f\t%6.3f\r\n',i,Pos(i,2),Pos(i,3),Pos(i,4),Pos(i,5));
end
fclose(fid);

%%
fid = fopen(sprintf('Si.Ge.%s.%d.xyz','shifted',sys),'w');
fprintf(fid,'%d\r\n',sys);
fprintf(fid,'%4.2f\t0\t0\t0\t%4.2f\t0\t0\t0\t%4.2f\r\n',max(Pos(:,3)),max(Pos(:,4))...
    ,max(Pos(:,5)));
for i = 1:sys
    if Pos(i,2)==1
        fprintf(fid,'Si\t%4.2f\t%4.2f\t%4.2f\r\n',Pos(i,3),Pos(i,4),Pos(i,5));
    elseif Pos(i,2)==2
        fprintf(fid,'Ge\t%4.2f\t%4.2f\t%4.2f\r\n',Pos(i,3),Pos(i,4),Pos(i,5));
    end
end
fclose(fid);






